### DISCUSSION TITLE
Update the torrc in GitHub Actions

### DISCUSSION THREAD

**Discussion by Lyro1** [community/community#24937] [Actions]
<p>Hello there,</p>
<p>I have a project in which I have to make requests via the tor service.<br>
I have no issue at all to install and start the tor service in my Github Action, but I need to change the <code>torrc</code> file, which is the configuration file for tor.</p>
<p>I tried to do so:</p>
<pre><code class="lang-auto">- name: Setup Tor config
      run: |
        sudo echo "ControlPort 9051" &gt;&gt; /etc/tor/torrc
        sudo echo "CookieAuthentication 1" &gt;&gt; /etc/tor/torrc
        sudo echo "CacheDirectoryGroupReadable 1" &gt;&gt; /etc/tor/torrc
</code></pre>
<p>But I get the following error:</p>
<pre><code class="lang-auto">Run sudo echo "ControlPort 9051" &gt;&gt; /etc/tor/torrc
9
/home/runner/work/_temp/d4dce9d5-4ca2-43a2-aa49-8a369f544a77.sh: line 1: /etc/tor/torrc: Permission denied
</code></pre>
<p>Is there any way I can modify my torrc for my Github Action to run?</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>This doesn’t work because of how shell redirection works: If you do</p>
<pre><code class="lang-bash">sudo echo "ControlPort 9051" &gt;&gt; /etc/tor/torrc
</code></pre>
<p>what happens is:</p>
<ul>
<li>The command <code>sudo echo "ControlPort 9051"</code> is executed</li>
<li>The output of that command is redirected and appended to <code>/etc/tor/torrc</code>
</li>
</ul>
<p>That redirect is done by the shell that executes your command, it is <em>not</em> part of the <code>sudo</code> command, so it runs with regular user rights. Only <code>echo</code> runs as root. :sweat_smile:</p>
<p>To make it work you need to make the process that writes to the file run as root. Two common solutions:</p>
<ol>
<li>Wrap everything in a root shell. You can do that by creating a script file and running that with sudo, or by wrapping the command with <code>bash -c</code>:</li>
</ol>
<pre><code class="lang-bash">sudo bash -c 'echo "ControlPort 9051" &gt;&gt; /etc/tor/torrc'
</code></pre>
<ol start="2">
<li>Send the echo output to a command that’s dedicated to writing to a file:</li>
</ol>
<pre><code class="lang-bash">echo "ControlPort 9051" | sudo tee --append /etc/tor/torrc
</code></pre>

**Comment by Simran-B** [comment]
<p>Re 2: Does <code>tee</code> run as root or how is it able to append to the file?</p>

**Comment by airtower-luna** [comment]
<p>I meant for it to, thank you for the catch! I’ve fixed the example. :sweat_smile:</p>

**Comment by Lyro1** [comment]
<p>Thanks, this worked as expected!</p>
<p>I faced another issue later on, but this will have its own new topic :slight_smile:</p>
